## Lab: SQL Practice

> Authors: Matt Brems & Jeff Hale

In this lab, you have 11 prompts. Most prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to help you practice your SQL - an **important** skill for working with data! 

Do everything with an SQL query in `pd.read_sql()`. 

If you want extra practice, do everything in pandas, too! 🐼

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The ERD for this database is here:

<img src="./images/erd.png" alt="erd" width="750"/>

In [1]:
# Imports

import pandas as pd
from sqlalchemy import create_engine

In [2]:
# Set up the engine to access the data.

# 'driver://username:password@host:port/database'
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [3]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.*
FROM stores as s;
"""

pd.read_sql(sql, engine)

store                         name store_status  \
0      2106  Hillstreet News and Tobacco            A   
1      2112               Mike's Liquors            I   
2      2113               Jamboree Foods            A   
3      2119           Manly Liquor Store            I   
4      2130         Sycamore Convenience            A   
...     ...                          ...          ...   
1968   9917       Honey Creek Distillery            A   
1969   9918  Paradise Distilling Company            A   
1970   9919            Dehner Distillery            A   
1971   9920        Louisiana Spirits LLC            A   
1972   9921                  maDIKwe USA            A   

                                          store_address address_info  
0     2217 College\nCedar Falls, IA 506130000\n(42.5...         None  
1     407 Sharp St.\nGlenwood, IA 515340000\n(41.046...         None  
2     1119 Market St. Box 71\nGowrie, IA 505430000\n...         None  
3     133 East Main\nManly, IA 504560000\n(43.286863...         None  
4     617 Sycamore\nWaterloo, IA 507030000\n(42.4978...         None  
...                                                 ...          ...  
1968  29538, 210th St.\nKeosauqua, IA 52565\n(40.740...         None  
1969  245, Railroad Ave\nDubuque, IA 52001\n(42.4873...         None  
1970  7500, University Ave\nClive, IA\n(41.600352803...         None  
1971  20909, South I-10 Frontage Rd\nLacassine, IA 7...         None  
1972  3101, Ingersoll Ave\nDes Moines, IA 50312\n(41...         None  

[1973 rows x 5 columns]

**Prompt 1 Answer**: 1973 rows

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth? It's okay that numbers and capital letters come before lower case letter.

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.*
FROM stores as s
ORDER BY s.name;
"""

pd.read_sql(sql, engine)


store                                name store_status  \
0      4986  218 Fuel Express & Chubby's Liquor            A   
1      3047          3047 Cub Foods / Iowa City            I   
2      3060               3060 Cub Foods / Ames            I   
3      3061         3061 Cub Foods / Sioux City            I   
4      4845                  3rd St Convenience            A   
...     ...                                 ...          ...   
1968   3992       Xo Food And Liquor / Waterloo            I   
1969   4066                       Xpress Liquor            I   
1970   4886                      Yr Dollar Stop            A   
1971   4908                Zapf's Pronto Market            A   
1972   4874                      Z's Quickbreak            A   

                                          store_address address_info  
0     68 Monroe St\nFloyd, IA 50435\n(43.12249790500...         None  
1     855 Highway 1 West\nIowa City, IA 522460000\n(...         None  
2     3121 Grand Ave\nAmes, IA 500100000\n(42.053458...         None  
3     1732 Hamilton Blvd\nSioux City, IA 511030000\n...         None  
4     216, 3rd St Se\nCedar Rapids, IA 52401\n(41.97...         None  
...                                                 ...          ...  
1968  428 Franklin St\nWaterloo, IA 507030000\n(42.5...         None  
1969  1201 North Jefferson Ste 100\nIndianola, IA 50...         None  
1970  1500 1st Ave Ne\nCedar Rapids, IA 52402\n(41.9...         None  
1971  107 1st St Sw\nElkader, IA 52043\n(42.85350191...         None  
1972  201, E Washington St\nMount Pleasant, IA 52641...         None  

[1973 rows x 5 columns]

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City

### Prompt 3

For sales where the category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what are the bottle price and number of bottles in each of the first 3 transactions?

In [9]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.category_name, s.vendor, s.btl_price, s.bottle_qty
FROM sales as s
WHERE s.category_name = 'IMPORTED VODKA'
ORDER BY s.vendor;
"""

pd.read_sql(sql, engine)


category_name                     vendor btl_price  bottle_qty
0       IMPORTED VODKA  Adamba Imports Int'l Inc.     $8.99          12
1       IMPORTED VODKA      A Hardy / U.S.A. Ltd.    $28.50           6
2       IMPORTED VODKA      A Hardy / U.S.A. Ltd.    $28.50           6
3       IMPORTED VODKA      A Hardy / U.S.A. Ltd.    $28.50           6
4       IMPORTED VODKA      A Hardy / U.S.A. Ltd.    $28.50           6
...                ...                        ...       ...         ...
124670  IMPORTED VODKA                       None    $16.33           3
124671  IMPORTED VODKA                       None    $16.33           2
124672  IMPORTED VODKA                       None    $16.33          12
124673  IMPORTED VODKA                       None    $16.33          12
124674  IMPORTED VODKA                       None    $16.33           2

[124675 rows x 4 columns]

**Prompt 3 Answer**: 
1. price: $8.99, qty: 12
2. price: $28.50, qty: 6
3. price: $28.50, qty: 6

This answer changes slightly every time I run the code

### Prompt 4

What is the total amount of liquor sold in February 2015? Display it in dollars and cents.

In [80]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.date AS date, SUM(s.total)
FROM sales as s
WHERE EXTRACT(MONTH from s.date) = '2'
AND EXTRACT(YEAR from s.date) = '2015'
GROUP BY date
"""

pd.read_sql(sql, engine)

date         sum
0  2015-02-02  1481821.78
1  2015-02-03  1109610.42
2  2015-02-04  1558001.57
3  2015-02-05  1246966.38
4  2015-02-06     1738.02
5  2015-02-09  1780048.14
6  2015-02-10  1338831.95
7  2015-02-11  1860931.81
8  2015-02-12  1009483.81
9  2015-02-16  1414626.83
10 2015-02-17  1147373.92
11 2015-02-18  1278745.71
12 2015-02-19  1063084.11
13 2015-02-23  1408529.36
14 2015-02-24  1169760.06
15 2015-02-25  1340126.96
16 2015-02-26  1085669.72

**Prompt 4 Answer**: The total amount of liquor sold in February 2015 

$ 21,295,350.55

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? 

Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. 

Rename the columns to make sure there isn't any confusion as to what they mean.

In [50]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(s.category_name) as count, s.vendor, SUM(s.bottle_qty) AS bottles, SUM(s.total)
FROM sales AS s
WHERE s.category_name = 'IMPORTED VODKA'
GROUP BY s.vendor
ORDER BY count DESC
LIMIT 1
"""

pd.read_sql(sql, engine)

count                            vendor  bottles         sum
0  42338  Pernod Ricard USA/Austin Nichols   403450  7880365.88

**Prompt 5 Answer**: 

Pernod Ricard: 42,338 transactions, 403,450 bottles and $ 7,880,365.88

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of money charged for a product over the cost to make that product.

In Iowa, the law states the minimum price at which liquor may be sold.

Create a column called `Markup` that shows the difference between the `btl_price` and the `state_btl_cost`.

In [52]:
# Use a SQL query to display the table discussed above.

sql = """
SELECT s.btl_price, s.state_btl_cost, s.btl_price - s.state_btl_cost AS Markup 
FROM sales AS s
"""

pd.read_sql(sql, engine)

btl_price state_btl_cost  markup
0          $17.24         $11.49   $5.75
1          $17.24         $11.49   $5.75
2          $17.24         $11.49   $5.75
3          $25.73         $17.15   $8.58
4           $7.50          $5.00   $2.50
...           ...            ...     ...
3049908    $17.24         $11.49   $5.75
3049909    $29.99         $19.99  $10.00
3049910    $17.24         $11.49   $5.75
3049911    $17.24         $11.49   $5.75
3049912    $29.99         $19.99  $10.00

[3049913 rows x 3 columns]

### Prompt 7

Calculate and display the "Percentage Markup" per store (just store id, not name) by dividing markup by the state bottle cost. 

This will tell you how much stores are increasing their average price over the state minimum.

In [62]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.store, AVG((s.btl_price / s.state_btl_cost) - 1) AS "Percentage Markup"
FROM sales AS s
GROUP BY s.store
"""

pd.read_sql(sql, engine)

store  Percentage Markup
0      2106           0.502086
1      2113           0.501741
2      2130           0.501469
3      2152           0.505025
4      2178           0.501654
...     ...                ...
1347   9013           0.500550
1348   9014           0.500496
1349   9018           0.500000
1350   9022           0.500395
1351   9023           0.500284

[1352 rows x 2 columns]

### Prompt 8

Calculate the average percentage markup per **store**. Return the average percentage markup and store number for the three stores with the largest markups.

In [60]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.store, AVG((s.btl_price / s.state_btl_cost) - 1) AS "Percentage Markup"
FROM sales AS s
GROUP BY s.store
ORDER BY "Percentage Markup" DESC
LIMIT 3
"""

pd.read_sql(sql, engine)

store  Percentage Markup
0   4013           0.525471
1   4024           0.521499
2   4266           0.518903

**Prompt 8 Answer**: The store with the highest average markup percentage is store 4013

### Prompt 9

Retrieve the three stores with the highest average markup percentage. Along with the store IDs and average markup percentage, return the minimum and maximum markup percentage for those stores, the names and addresses of those store.

What are the names of the 3 stores?

In [70]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT st.name AS store_name, st.store_address, st.store, AVG((s.btl_price / s.state_btl_cost) - 1) AS "Percentage Markup", MIN((s.btl_price / s.state_btl_cost) - 1) AS "Min. Percentage Markup", MAX((s.btl_price / s.state_btl_cost) - 1) AS "Max. Percentage Markup"
FROM sales AS s
INNER JOIN stores AS st ON st.store = s.store
GROUP BY st.store
ORDER BY "Percentage Markup" DESC
LIMIT 3
"""

pd.read_sql(sql, engine)

store_name  \
0      Wal-Mart 0841 / Tipton   
1  Wal-Mart 1546 / Iowa Falls   
2  Wal-Mart 1683 / Shenandoah   

                                       store_address  store  \
0  1126 Highway 38 North\nTipton, IA 527720000\n(...   4013   
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...   4024   
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...   4266   

   Percentage Markup  Min. Percentage Markup  Max. Percentage Markup  
0           0.525471                0.498571                6.708763  
1           0.521499                0.498462                6.708763  
2           0.518903                0.498462                6.708763

**Prompt 9 Answer**: The three stores with the highest markups are:
1. Wal-Mart 0841 / Tipton
2. Wal-Mart 1546 / Iowa Falls
3. Wal-Mart 1683 / Shenandoah

### Prompt 10

Which nine counties sell the highest number of bottles of liquor per capita? Calculate the average bottle size of the bottles of liquor sold by these counties, rounded to the nearest first decimal place. Also display the county name, and average bottle size.

Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

In [78]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT c.county, SUM(s.bottle_qty) / c.population AS "Bottles Sold per Capita", ROUND(AVG(s.liter_size), 1) AS avg_size
FROM sales as s
INNER JOIN counties as c ON c.county = s.county
GROUP BY c.county
ORDER BY "Bottles Sold per Capita" DESC
LIMIT 9
"""

pd.read_sql(sql, engine)

county  Bottles Sold per Capita  avg_size
0      Dickinson                       22     974.2
1           Polk                       15     891.9
2     Black Hawk                       15     861.0
3    Cerro Gordo                       14     986.4
4          Scott                       13     869.9
5        Johnson                       13     895.2
6           Linn                       12     885.1
7  Pottawattamie                       11     890.9
8        Kossuth                       11    1040.5

**Prompt 10 Answer**: 

1. Dickinson
2. Polk
3. Black Hawk
4. Cerro Gordo
5. Scott
6. Johnson
7. Linn
8. Pottawattamie
9. Kossuth

Kossuth County sells the largest bottle size on average.

### Prompt 11

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions, if any, have a value of `total` that is not equal to `btl_price` times `bottle_qty`?

> You might get a type error. We cannot compare type `money` to type `real`. You may have to convert `total` to type `money` to compare the two directly.

In [93]:
# Use a SQL query to find the answer to the above prompt.

# GET all totals
sql = """
SELECT s.btl_price, s.bottle_qty, CAST(s.total AS money) AS total, s.btl_price * s.bottle_qty AS "Check"
FROM sales AS s

"""


pd.read_sql(sql, engine)

btl_price  bottle_qty    total    Check
0          $17.24           4   $68.96   $68.96
1          $17.24          12  $206.88  $206.88
2          $17.24          12  $206.88  $206.88
3          $25.73           6  $154.38  $154.38
4           $7.50           6   $45.00   $45.00
...           ...         ...      ...      ...
3049908    $17.24          12  $206.88  $206.88
3049909    $29.99          30  $899.70  $899.70
3049910    $17.24          12  $206.88  $206.88
3049911    $17.24          12  $206.88  $206.88
3049912    $29.99          12  $359.88  $359.88

[3049913 rows x 4 columns]

In [94]:
# Use a SQL query to find the answer to the above prompt.

# GET all totals
sql = """
SELECT s.btl_price, s.bottle_qty, CAST(s.total AS money) AS total, s.btl_price * s.bottle_qty AS "Check"
FROM sales AS s
WHERE CAST(s.total AS money) != s.btl_price * s.bottle_qty
"""


pd.read_sql(sql, engine)

Empty DataFrame
Columns: [btl_price, bottle_qty, total, Check]
Index: []

**Prompt 11 Answer**: 

Apparently there are no transactions with totals different to bottle quantity times bottle price